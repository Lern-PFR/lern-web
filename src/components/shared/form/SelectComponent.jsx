import styled from 'styled-components';
import {
	border,
	flexbox,
	layout,
	margin,
	padding,
} from 'styled-system';
import PropTypes from 'prop-types';
import ReactSelect from 'react-select';
import { forwardRef, useRef, useMemo } from 'react';
import { errorSelectStyle, selectStyle } from 'theme/formStyles';
import { primary, crimson, jasmine } from 'theme/colors';
import { getTypographyStyleByName } from '../typography';

const StyledSelect = styled(ReactSelect)(
	border,
	flexbox,
	layout,
	margin,
	padding,
);

/**
 * @name SelectComponent
 * @description A component used to display a styled select element.
 *
 * @author Christopher Walker
 *
 * @param {string|number}	id					The id of the select element.
 * @param {array}			options				The array of options to display.
 * @param {array}			[selectedOptions]	The array of pre-selected options.
 * @param {string}			[textStyle]			The text style for this select.
 * @param {bool}			[disabled]			Whether or not the element is disabled.
 * @param {bool}			[required]			Whether or not the element is required.
 * @param {bool}			[hasError]			Whether or not the select contains a validation error.
 * @param {string|number}	[placeholder]		The placeholder text to display.
 * @param {bool}			[multiple]			Whether or not multiple options can be selected.
 * @param {string}			[noOptionsMessage]	Text to display when no options were found.
 */
const SelectComponent = forwardRef(
	({ id, disabled, required, multiple, placeholder, options, selectedOptions, hasError, textStyle, noOptionsMessage, ...otherProps }, ref) => {
		const defaultRef = useRef();
		const resolvedRef = ref || defaultRef;
		const typographyStyle = useMemo(() => getTypographyStyleByName(textStyle), [textStyle]);

		return (
			<StyledSelect
				noOptionsMessage={() => noOptionsMessage}
				isClearable={!required}
				isDisabled={disabled}
				defaultValue={selectedOptions || ''}
				isMulti={multiple}
				placeholder={placeholder}
				id={id}
				options={options}
				ref={resolvedRef}
				styles={hasError && !disabled ? errorSelectStyle : selectStyle}
				theme={(theme) => ({
					...theme,
					colors: {
						...theme.colors,
						primary: hasError ? crimson.default : primary.default,
						primary25: jasmine.darker1,
						primary50: primary.lighter1,
					},
				})}
				{...typographyStyle}
				{...otherProps}
			/>
		);
	}
);

SelectComponent.displayName = 'SelectComponent';

SelectComponent.propTypes = {
	id: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
	]).isRequired,
	textStyle: PropTypes.string,
	disabled: PropTypes.bool,
	required: PropTypes.bool,
	hasError: PropTypes.bool,
	multiple: PropTypes.bool,
	placeholder: PropTypes.string,
	options: PropTypes.arrayOf(
		PropTypes.shape({
			value: PropTypes.oneOfType([
				PropTypes.string,
				PropTypes.number,
			]).isRequired,
			label: PropTypes.oneOfType([
				PropTypes.string,
				PropTypes.number,
				PropTypes.node,
			]).isRequired,
		}),
	).isRequired,
	selectedOptions: PropTypes.oneOfType([
		PropTypes.arrayOf(
			PropTypes.shape({
				value: PropTypes.oneOfType([
					PropTypes.string,
					PropTypes.number,
				]).isRequired,
				label: PropTypes.oneOfType([
					PropTypes.string,
					PropTypes.number,
					PropTypes.node,
				]).isRequired,
			}),
		),
		PropTypes.shape({
			value: PropTypes.oneOfType([
				PropTypes.string,
				PropTypes.number,
			]).isRequired,
			label: PropTypes.oneOfType([
				PropTypes.string,
				PropTypes.number,
				PropTypes.node,
			]).isRequired,
		}),
	]),
	noOptionsMessage: PropTypes.string,
};

SelectComponent.defaultProps = {
	textStyle: 'brevier',
	disabled: false,
	required: false,
	hasError: false,
	multiple: false,
	placeholder: ' - ',
	selectedOptions: [],
	noOptionsMessage: 'No options',
};

export default SelectComponent;
