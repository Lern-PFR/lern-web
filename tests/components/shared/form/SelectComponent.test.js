import { shallow } from 'enzyme';
import SelectComponent from 'components/shared/form/SelectComponent';
import { cleanup } from '@testing-library/react';

describe('SelectComponent', () => {
	afterEach(() => {
		cleanup();
	});

	const options = [
		{
			value: 1,
			label: 'One',
		},
		{
			value: '2',
			label: 'Two',
		},
		{
			value: 3,
			label: 'Three',
		},
		{
			value: '4',
			label: 'Four',
		},
	];

	it('should match previous snapshot', () => {
		const sut = (<SelectComponent id="select" options={options} />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for disabled state', () => {
		const sut = (<SelectComponent disabled id="select" options={options} />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for select with custom placeholder', () => {
		const sut = (<SelectComponent id="select" placeholder="Select an option" options={options} />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for select with validation error', () => {
		const sut = (<SelectComponent id="select" hasError options={options} />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for select with custom noOptionsMessage', () => {
		const sut = (<SelectComponent id="select" options={[]} noOptionsMessage="Sorry, we couldn't find any options" />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for select with pre-selected option', () => {
		const sut = (<SelectComponent id="select" selectedOptions={options[1]} options={options} />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for multiple option select', () => {
		const sut = (<SelectComponent id="select" multiple options={options} />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for multiple option select with pre-selected options', () => {
		const sut = (<SelectComponent id="select" multiple options={options} selectedOptions={[options[0], options[2]]} />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});
});
