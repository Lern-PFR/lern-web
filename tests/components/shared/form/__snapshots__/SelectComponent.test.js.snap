// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectComponent should match previous snapshot 1`] = `
<Styled(StateManager)
  defaultValue={Array []}
  fontFamily="IBM Plex Sans, Helvetica Neue, Arial, sans-serif"
  fontSize=".778rem"
  fontWeight={400}
  id="select"
  isClearable={true}
  isDisabled={false}
  isMulti={false}
  letterSpacing={0.32}
  lineHeight="1.125em"
  noOptionsMessage={[Function]}
  options={
    Array [
      Object {
        "label": "One",
        "value": 1,
      },
      Object {
        "label": "Two",
        "value": "2",
      },
      Object {
        "label": "Three",
        "value": 3,
      },
      Object {
        "label": "Four",
        "value": "4",
      },
    ]
  }
  placeholder=" - "
  styles={
    Object {
      "control": [Function],
      "menu": [Function],
      "multiValueLabel": [Function],
      "option": [Function],
      "placeholder": [Function],
      "valueContainer": [Function],
    }
  }
  theme={[Function]}
/>
`;

exports[`SelectComponent should match previous snapshot for disabled state 1`] = `
<Styled(StateManager)
  defaultValue={Array []}
  fontFamily="IBM Plex Sans, Helvetica Neue, Arial, sans-serif"
  fontSize=".778rem"
  fontWeight={400}
  id="select"
  isClearable={true}
  isDisabled={true}
  isMulti={false}
  letterSpacing={0.32}
  lineHeight="1.125em"
  noOptionsMessage={[Function]}
  options={
    Array [
      Object {
        "label": "One",
        "value": 1,
      },
      Object {
        "label": "Two",
        "value": "2",
      },
      Object {
        "label": "Three",
        "value": 3,
      },
      Object {
        "label": "Four",
        "value": "4",
      },
    ]
  }
  placeholder=" - "
  styles={
    Object {
      "control": [Function],
      "menu": [Function],
      "multiValueLabel": [Function],
      "option": [Function],
      "placeholder": [Function],
      "valueContainer": [Function],
    }
  }
  theme={[Function]}
/>
`;

exports[`SelectComponent should match previous snapshot for multiple option select 1`] = `
<Styled(StateManager)
  defaultValue={Array []}
  fontFamily="IBM Plex Sans, Helvetica Neue, Arial, sans-serif"
  fontSize=".778rem"
  fontWeight={400}
  id="select"
  isClearable={true}
  isDisabled={false}
  isMulti={true}
  letterSpacing={0.32}
  lineHeight="1.125em"
  noOptionsMessage={[Function]}
  options={
    Array [
      Object {
        "label": "One",
        "value": 1,
      },
      Object {
        "label": "Two",
        "value": "2",
      },
      Object {
        "label": "Three",
        "value": 3,
      },
      Object {
        "label": "Four",
        "value": "4",
      },
    ]
  }
  placeholder=" - "
  styles={
    Object {
      "control": [Function],
      "menu": [Function],
      "multiValueLabel": [Function],
      "option": [Function],
      "placeholder": [Function],
      "valueContainer": [Function],
    }
  }
  theme={[Function]}
/>
`;

exports[`SelectComponent should match previous snapshot for multiple option select with pre-selected options 1`] = `
<Styled(StateManager)
  defaultValue={
    Array [
      Object {
        "label": "One",
        "value": 1,
      },
      Object {
        "label": "Three",
        "value": 3,
      },
    ]
  }
  fontFamily="IBM Plex Sans, Helvetica Neue, Arial, sans-serif"
  fontSize=".778rem"
  fontWeight={400}
  id="select"
  isClearable={true}
  isDisabled={false}
  isMulti={true}
  letterSpacing={0.32}
  lineHeight="1.125em"
  noOptionsMessage={[Function]}
  options={
    Array [
      Object {
        "label": "One",
        "value": 1,
      },
      Object {
        "label": "Two",
        "value": "2",
      },
      Object {
        "label": "Three",
        "value": 3,
      },
      Object {
        "label": "Four",
        "value": "4",
      },
    ]
  }
  placeholder=" - "
  styles={
    Object {
      "control": [Function],
      "menu": [Function],
      "multiValueLabel": [Function],
      "option": [Function],
      "placeholder": [Function],
      "valueContainer": [Function],
    }
  }
  theme={[Function]}
/>
`;

exports[`SelectComponent should match previous snapshot for select with custom noOptionsMessage 1`] = `
<Styled(StateManager)
  defaultValue={Array []}
  fontFamily="IBM Plex Sans, Helvetica Neue, Arial, sans-serif"
  fontSize=".778rem"
  fontWeight={400}
  id="select"
  isClearable={true}
  isDisabled={false}
  isMulti={false}
  letterSpacing={0.32}
  lineHeight="1.125em"
  noOptionsMessage={[Function]}
  options={Array []}
  placeholder=" - "
  styles={
    Object {
      "control": [Function],
      "menu": [Function],
      "multiValueLabel": [Function],
      "option": [Function],
      "placeholder": [Function],
      "valueContainer": [Function],
    }
  }
  theme={[Function]}
/>
`;

exports[`SelectComponent should match previous snapshot for select with custom placeholder 1`] = `
<Styled(StateManager)
  defaultValue={Array []}
  fontFamily="IBM Plex Sans, Helvetica Neue, Arial, sans-serif"
  fontSize=".778rem"
  fontWeight={400}
  id="select"
  isClearable={true}
  isDisabled={false}
  isMulti={false}
  letterSpacing={0.32}
  lineHeight="1.125em"
  noOptionsMessage={[Function]}
  options={
    Array [
      Object {
        "label": "One",
        "value": 1,
      },
      Object {
        "label": "Two",
        "value": "2",
      },
      Object {
        "label": "Three",
        "value": 3,
      },
      Object {
        "label": "Four",
        "value": "4",
      },
    ]
  }
  placeholder="Select an option"
  styles={
    Object {
      "control": [Function],
      "menu": [Function],
      "multiValueLabel": [Function],
      "option": [Function],
      "placeholder": [Function],
      "valueContainer": [Function],
    }
  }
  theme={[Function]}
/>
`;

exports[`SelectComponent should match previous snapshot for select with pre-selected option 1`] = `
<Styled(StateManager)
  defaultValue={
    Object {
      "label": "Two",
      "value": "2",
    }
  }
  fontFamily="IBM Plex Sans, Helvetica Neue, Arial, sans-serif"
  fontSize=".778rem"
  fontWeight={400}
  id="select"
  isClearable={true}
  isDisabled={false}
  isMulti={false}
  letterSpacing={0.32}
  lineHeight="1.125em"
  noOptionsMessage={[Function]}
  options={
    Array [
      Object {
        "label": "One",
        "value": 1,
      },
      Object {
        "label": "Two",
        "value": "2",
      },
      Object {
        "label": "Three",
        "value": 3,
      },
      Object {
        "label": "Four",
        "value": "4",
      },
    ]
  }
  placeholder=" - "
  styles={
    Object {
      "control": [Function],
      "menu": [Function],
      "multiValueLabel": [Function],
      "option": [Function],
      "placeholder": [Function],
      "valueContainer": [Function],
    }
  }
  theme={[Function]}
/>
`;

exports[`SelectComponent should match previous snapshot for select with validation error 1`] = `
<Styled(StateManager)
  defaultValue={Array []}
  fontFamily="IBM Plex Sans, Helvetica Neue, Arial, sans-serif"
  fontSize=".778rem"
  fontWeight={400}
  id="select"
  isClearable={true}
  isDisabled={false}
  isMulti={false}
  letterSpacing={0.32}
  lineHeight="1.125em"
  noOptionsMessage={[Function]}
  options={
    Array [
      Object {
        "label": "One",
        "value": 1,
      },
      Object {
        "label": "Two",
        "value": "2",
      },
      Object {
        "label": "Three",
        "value": 3,
      },
      Object {
        "label": "Four",
        "value": "4",
      },
    ]
  }
  placeholder=" - "
  styles={
    Object {
      "control": [Function],
      "menu": [Function],
      "multiValueLabel": [Function],
      "option": [Function],
      "placeholder": [Function],
      "valueContainer": [Function],
    }
  }
  theme={[Function]}
/>
`;
